### ISSUE TITLE
P2P Sync

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#281]
User can sync data to another device via bluetooth 	Users can sync data to other CHWs or supervisors

**Comment by f-odhiambo** [comment]
This should cover cases such as battery drain, failure on connectivity, dirty read, sync failure, restore to last best state

**Comment by bausmeier** [comment]
Based on the approach outlined in the related [discussion](https://github.com/opensrp/fhircore/discussions/691) and [doc](https://docs.google.com/document/d/1R5GgzpkhnFkEdIcqEu3FjL9mYHdnTATni8EPYdHOyyI/edit?usp=sharing), this is the set of tasks which Trevor and I will be working on:

- [ ] Create `SyncPayload` class.
- [ ] Create `SyncSenderSession` interface.
- [ ] Create `SyncReceiverSession` interface.
- [ ] Create `ConflictResolutionStrategy` interface.
- [ ] Implement `SyncSender` class.
- [ ] Implement `OursConflictResolutionStrategy` class.
- [ ] Implement `TheirsConflictResolutionStrategy` class.
- [ ] Implement `SyncRecevier` class.
- [ ] Implement `SocketSyncSenderSession` class.
- [ ] Implement `SocketSyncReceiverSession` class.
- [ ] Implement Wi-Fi Direct listener interfaces (https://developer.android.com/guide/topics/connectivity/wifip2p#api-overview).
- [ ] Implement sync UI components.

**Comment by bausmeier** [comment]
Collating some of the relevant links:

GitHub issues and discussions:
- https://github.com/google/android-fhir/issues/258
- https://github.com/google/android-fhir/issues/275

Android FHIR SDK server sync implementation:
- https://github.com/google/android-fhir (See [the `Sync` class](https://github.com/google/android-fhir/blob/d3c3feda998e7c7fdc20c07b539247c892c0fb72/engine/src/main/java/com/google/android/fhir/sync/Sync.kt))
- https://github.com/opensrp/fhircore

OpenSRP Android P2P Sync:
- https://github.com/OpenSRP/android-p2p-sync

GSoC findings and demos:
- https://docs.google.com/document/d/1o9m1vUh2tkAuzKS-JIFEohUHeVWeas36ob0ztOlBF0A/edit
- https://gist.github.com/maanuanubhav999/e6b2899dd308d35320a90ff215f312d7
- https://github.com/ekigamba/wifi-direct-sample
- https://github.com/opensrp/labs/tree/main/GSOC/2021/fhir-peer-to-peer

Past meeting notes:
- https://docs.google.com/document/d/1rdRu0_z3KN7fUUoqq9ZuVsUdhjtf70p13axwjT7ZRVU/edit

**Comment by bausmeier** [comment]
@f-odhiambo On one of the calls the other day we talked about how the permissions part of this issue is a large and complex piece of work which should rather be tackled on its own. Could we split that part out into a new issue so that this one focuses only on the P2P sync part? Or maybe we could just note in the description that it's out of scope for this piece?

**Comment by f-odhiambo** [comment]
> @f-odhiambo On one of the calls the other day we talked about how the permissions part of this issue is a large and complex piece of work which should rather be tackled on its own. Could we split that part out into a new issue so that this one focuses only on the P2P sync part? Or maybe we could just note in the description that it's out of scope for this piece?

@bausmeier I think I can note andalso create a Discussion board around it  . Ill follow up on this

**Comment by pld** [comment]
For the UI components, let's discuss w/Roger before we start any work on this, I think we'll want to have him design it out a bit first

**Comment by bausmeier** [comment]
Will do. I think we'll need input from some others on the UI aspect, and how best to implement it to. I'm planning on setting up some time to chat about this.

**Comment by jingtang10** [comment]
Discussed with @bausmeier this morning and draw this diagram to represent the data flow:

<img width="959" alt="Screen Shot 2021-11-29 at 13 03 55" src="https://user-images.githubusercontent.com/28757340/143873099-67efd465-f50e-4fc8-88a9-bbaf6901f7a1.png">

@f-odhiambo @pld does this represent the use case correctly?

**Comment by pld** [comment]
My understanding is that this represents the most complex scenario we've encountered so far. We also have cases with no DHO, i.e. the CHSSs have connectivity and connect directly to the server.

@AnnieMungai what types of configurations do you think we'll see in the Liberia project?

@rowo does this cover the campaigns use-case, or are there other use-cases we'll need to support in the next year you can think of that this does not cover?

**Comment by jingtang10** [comment]
Thanks @pld. The most important thing for the SDK support here is that there will be 2 types of nodes
1. CHWs and CHSSs where all changes are synced via P2P
2. DHOs (or the CHSSs when there is no DHOs) where it has P2P with downstream nodes but also uses FHIR RESTful API to sync with server

**Comment by AnnieMungai** [comment]
@pld currently any user can be able to sync data from any other user as above mentioned and also be able to sync data online. 
CHA >> CHA >> CHSS >> DHO >> Server
CHA >> CHSS >> DHO >> Server
CHA >> CHSS >> CHSS>> Server
CHSS >> CHSS >> Server
CHSS >> Server
CHA >> Server (this is rare but happens for users who have a rare connection to the internet)

**Comment by bausmeier** [comment]
@jingtang10 Does any user being able to sync to the server invalidate the approach we had in mind? I think that 1-5 should still be workable individually, but the combination of CHAs being able to sync with one another (1) and also being able to sync to the server (6) might be a problem. I can imagine that CHSS >> CHSS could also complicate things quite a lot.
